classes/field_controller.php

```diff
@@ -151,6 +158,28 @@ public function config_form_validation(array $data, $files = array()) : array {
         } catch (\Exception $e) {
             $err['configdata[dynamicsql]'] = get_string('sqlerror', 'customfield_dynamic') . ': ' .$e->getMessage();
         }
+
+        try {
+            // Ensure the datafield is valid.
+            $newdatafield = $data['configdata']['datafield'];
+            if (!\core_customfield\data::has_property($newdatafield)) {
+                $err['configdata[datafield]'] = get_string('datafieldinvalid', 'customfield_dynamic', $newdatafield);
+            }
+
+            // Check if a new datafield is set and a migration is needed.
+            $prevdatafield = $this->get_configdata_property('datafield');
+            $migrate = $newdatafield !== $prevdatafield && $newdatafield !== 'value';
+
+            // If we have no errors and the field already exists, attempt to migrate values to match.
+            if (empty($err) && !empty($this->get('id')) && $migrate) {
+                $sql = "UPDATE {customfield_data} SET $newdatafield = value WHERE fieldid = :fieldid";
```

Comment:
should this be SET $newdatafield = $prevdatafield ?

like edge cases we don't strictly need right now but if we set it to charvalue and then swap it to value then it should migrate both directions

Reply:
The default behavior of the data controller class is to save the value to both the specified value (i.e. charvalue) and value, so value should contain the relevant data.

https://github.com/moodle/moodle/blob/main/customfield/classes/data_controller.php#L204